src/Catel.MVVM/Catel.MVVM.Shared/Services/Extensions/IUIVisualizerServiceExtensions.cs

```diff
@@ -188,6 +191,83 @@ private static IViewModelFactory GetViewModelFactory(IUIVisualizerService uiVisu
+
+            var viewLocator = uiVisualizerService.GetDependencyResolver().Resolve<IViewLocator>();
+            var viewType = viewLocator.ResolveView(viewModel.GetType());
+            var viewManager = uiVisualizerService.GetServiceLocator().ResolveType<IViewManager>();
+            var view = viewManager.GetFirstOrDefaultInstance(viewType);
+            if (view == null)
+            {
+                return uiVisualizerService.Show(typeof(TViewModel).FullName, null, completedProc);
+            }
+
+            var window = view as System.Windows.Window;
+            if (window == null)
+            {
+                return false;
+            }
+            return window.Activate();
+        }
+
+        /// <summary>
+        /// Creates a window in non-modal state. If a window with the specified viewModelType exists, the window is activated instead of being created.
+        /// </summary>
+        /// <param name="uiVisualizerService">The UI visualizer service.</param>
+        /// <typeparam name="TViewModel">The type of the view model.</typeparam>
+        /// <param name="model">The model to be injected into the view model, can be <c>null</c>.</param>
+        /// <param name="completedProc">The completed proc. Not applicable if window already exists.</param>
+        /// <returns><c>true</c> if shown or activated successfully, <c>false</c> otherwise.</returns>
+        /// <exception cref="ArgumentNullException">The <paramref name="uiVisualizerService" /> is <c>null</c>.</exception>
+        public static Task<bool?> ShowOrActivateAsync<TViewModel>(this IUIVisualizerService uiVisualizerService, object model = null, EventHandler<UICompletedEventArgs> completedProc = null) 
+        {
+            Argument.IsNotNull("uiVisualizerService", uiVisualizerService);
+
+            var viewModelManager = uiVisualizerService.GetServiceLocator().ResolveType<IViewModelManager>();
+            var viewModel = viewModelManager.GetFirstOrDefaultInstance(typeof(TViewModel));
+            if (viewModel == null)
+            {
+                var viewModelFactory = GetViewModelFactory(uiVisualizerService);
+                var vm = viewModelFactory.CreateViewModel(typeof(TViewModel), model);
+                return uiVisualizerService.ShowAsync(vm, completedProc);
+            }
+
+            var viewLocator = uiVisualizerService.GetDependencyResolver().Resolve<IViewLocator>();
+            var viewType = viewLocator.ResolveView(viewModel.GetType());
+            var viewManager = uiVisualizerService.GetServiceLocator().ResolveType<IViewManager>();
+            var view = viewManager.GetFirstOrDefaultInstance(viewType);
+            if (view == null)
+            {
+                return uiVisualizerService.ShowAsync(typeof(TViewModel).FullName, null, completedProc);
+            }
+
+            var window = view as System.Windows.Window;
+            if (window == null)
+            {
+                return new Task<bool?>(() => false);
```

Comment:
Use return Task.FromResult(false) (is cached instance, so faster and saves a garbage collection object).